s3/deployment/src/main/java/io/quarkus/amazon/s3/deployment/DevServicesLocalStackProcessor.java

```diff
@@ -0,0 +1,213 @@
+package io.quarkus.amazon.s3.deployment;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.time.Duration;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+
+import org.jboss.logging.Logger;
+import org.testcontainers.containers.localstack.LocalStackContainer;
+import org.testcontainers.utility.DockerImageName;
+
+import io.quarkus.amazon.s3.runtime.S3BuildTimeConfig;
+import io.quarkus.amazon.s3.runtime.S3DevServicesBuildTimeConfig;
+import io.quarkus.deployment.IsDockerWorking;
+import io.quarkus.deployment.IsNormal;
+import io.quarkus.deployment.annotations.BuildStep;
+import io.quarkus.deployment.builditem.CuratedApplicationShutdownBuildItem;
+import io.quarkus.deployment.builditem.DevServicesResultBuildItem;
+import io.quarkus.deployment.builditem.DevServicesResultBuildItem.RunningDevService;
+import io.quarkus.deployment.builditem.DevServicesSharedNetworkBuildItem;
+import io.quarkus.deployment.builditem.LaunchModeBuildItem;
+import io.quarkus.deployment.console.ConsoleInstalledBuildItem;
+import io.quarkus.deployment.console.StartupLogCompressor;
+import io.quarkus.deployment.dev.devservices.GlobalDevServicesConfig;
+import io.quarkus.deployment.logging.LoggingSetupBuildItem;
+import io.quarkus.devservices.common.ConfigureUtil;
+import io.quarkus.devservices.common.ContainerAddress;
+import io.quarkus.devservices.common.ContainerLocator;
+import software.amazon.awssdk.auth.credentials.AwsBasicCredentials;
+import software.amazon.awssdk.auth.credentials.StaticCredentialsProvider;
+import software.amazon.awssdk.regions.Region;
+import software.amazon.awssdk.services.s3.S3Client;
+import software.amazon.awssdk.services.s3.model.CreateBucketRequest;
+
+/**
+ * Starts a local stack container to emulate AWS services
+ */
+public class DevServicesLocalStackProcessor {
```

Comment:
Shouldn't the localstack container be started once for all the services? That's the approach we have in the tests and I think it's probably better.
Not saying that we cannot merge this as a first step but my guess is that we need a more general solution if we want to generalize this.

Reply:
It gets complex with label based service sharing, e.g. one app just wants s3, but another wants s3 + something else, it seemed simpler to just start a single container for now.

Reply:
@stuartwdouglas what do you mean by label based services sharing ? Should one localstack be started with all requested services ?

I guess a pattern like the datasource services may help. A set of AmazonServiceBuildItem grab by a AmazonDevServicesProcessor that will start the LocalStackContainer with the requested Services and let the buildItem prepare the stack and return map of configuration properties. Then, have a deployment project for each services to provide such AmazonServiceBuildItem.

Reply:
https://github.com/scrocquesel/quarkus-amazon-services/commit/98ee25549b73cd410989580c9b53a668177a08f1

I go straight to something that work. But it still need some work around configuration

Reply:
I have added dynamodb support and update integration tests to use devservices.
````
 Amazon Dev Services for localstack (dynamodb, s3) started.
````

@gsmet is this you were thinking about ?

As testcontainers, is pushing all the conf as env vars, I guess it could be possible to read them from a matched `container.config.env` to guess which services are enabled and compare with requested services. The idea is that if you need a group of applications to share a common localstack, it must be the app with the most services that is started first. Then, other apps will discover the container with the same service name label. The processor will check that the env var of the container is compatible or will warn the user if not. If one application is using another aws service that is not to be shared, then you can use a different service name and it will trigger another container.

- App1 requires dynamodb, s3, sqs
- App2 requires sqs

You could have only one localstack but App1 should always be started first so that it can share sqs with App2 on the same pool. If App2 is started first, when App1 is started, dev services will detect that dynamodb and s3 is missing and will fail. Or you can have a localstack for sqs and one for dynamodb and s3.